### PR TITLE
[MODORDSTOR-458]. Invalid reference is displayed in PO line after changing instance connection for the P/E mix order

### PR DESCRIPTION
**Pull request by BKadirkhodjaev**
## Purpose

- <https://folio-org.atlassian.net/browse/MODORDSTOR-458>

## Approach

- Refactor holding patch operation to avoid retrieving a holding that we just deleted with the "Delete Holdings" option

### CODE REVIEW

File: src/main/java/org/folio/models/orders/lines/update/OrderLineUpdateInstanceHolder.java [hunk 1]
```diff
@@ -9,9 +10,11 @@
 import org.folio.rest.jaxrs.model.PoLine;
 
 public class OrderLineUpdateInstanceHolder {
+
   private PoLine storagePoLine;
   private PatchOrderLineRequest patchOrderLineRequest;
   private StoragePatchOrderLineRequest storagePatchOrderLineRequest;
+  private final List<String> deletedHoldingIds = new ArrayList<>();
```
Comment (TarasSpashchenko) [inline src/main/java/org/folio/models/orders/lines/update/OrderLineUpdateInstanceHolder.java:17]: **[Suggestion] Documentation Enhancement**

Consider adding Javadoc to the new `deletedHoldingIds` field and the `addDeletedHoldingId` method to explain their purpose in preventing "Invalid reference" issues.

```java
/**
 * List of holding IDs that have been deleted during the instance update process.
 * Used to prevent processing of holdings that were already deleted via "Delete Holdings" operation.
 */
private final List<String> deletedHoldingIds = new ArrayList<>();

/**
 * Adds a holding ID to the list of deleted holdings.
 * @param holdingId the ID of the holding that was deleted
 */
public void addDeletedHoldingId(String holdingId) {
    this.deletedHoldingIds.add(holdingId);
}
```

File: src/main/java/org/folio/service/inventory/InventoryHoldingManager.java [hunk 2]
```diff
@@ -253,6 +252,9 @@ public Future<String> getOrCreateHoldingRecordByInstanceAndLocation(String insta
     if (Objects.isNull(location.getLocationId())) {
       return getHoldingById(location.getHoldingId(), true, requestContext)
         .compose(holding -> {
+          if (Objects.isNull(holding)) {
```
Comment (TarasSpashchenko) [inline src/main/java/org/folio/service/inventory/InventoryHoldingManager.java:255]: **[Suggestion] Logging Enhancement**

The null check is essential and correctly addresses the root cause of the issue. Consider adding a debug log statement to help with troubleshooting:

```java
if (Objects.isNull(holding)) {
    logger.debug("getOrCreateHoldingRecordByInstanceAndLocation:: holding not found for holdingId: {}, skipping processing", location.getHoldingId());
    return Future.succeededFuture();
}
```

This would help track when holdings have been deleted and processing is skipped.

File: src/main/java/org/folio/service/orders/lines/update/instance/BaseOrderLineUpdateInstanceStrategy.java [hunk 3]
```diff
@@ -40,21 +40,22 @@ public Future<Void> updateInstance(OrderLineUpdateInstanceHolder holder, Request
...
-            return deleteHoldingWithoutItems(location.getHoldingId(), locationContext)
-              .onSuccess(v -> logger.info("deleteAbandonedHoldings:: operation succeeded for holdingId: {}", location.getHoldingId()))
-              .onFailure(e -> logger.error("Failed to delete abandoned holdings for holdingId: {}", location.getHoldingId(), e));
-          }).toList()
-      )
+      poLine.getLocations()
+        .stream()
+        .filter(location -> StringUtils.isNotEmpty(location.getHoldingId()))
+        .map(location -> {
+          var locationContext = RequestContextUtil.createContextWithNewTenantId(requestContext, location.getTenantId());
```
Comment (TarasSpashchenko) [inline src/main/java/org/folio/service/orders/lines/update/instance/BaseOrderLineUpdateInstanceStrategy.java:53]: **[Suggestion] Method Architecture**

Good architectural improvement! The method signature change to accept `OrderLineUpdateInstanceHolder` and track deleted holdings is well-designed. The holder pattern centralizes state management and ensures deleted holdings are properly tracked across the deletion process.

The logging statements provide good visibility into the deletion operations, which is important for this critical workflow.

File: src/main/java/org/folio/service/orders/lines/update/instance/WithHoldingOrderLineUpdateInstanceStrategy.java [hunk 4]
```diff
@@ -160,8 +157,16 @@ private Future<Void> findOrCreateHoldingsAndUpdateItems(OrderLineUpdateInstanceH
     logger.info("findOrCreateHoldingsAndUpdateItems:: start processing for new instanceId: {}, holdingId: {}, tenantId(for ECS): {}",
       newInstanceId, holdingId, location.getTenantId());
     return inventoryInstanceManager.createShadowInstanceIfNeeded(newInstanceId, locationContext)
-      .compose(instance -> inventoryHoldingManager.getOrCreateHoldingRecordByInstanceAndLocation(newInstanceId, location, locationContext))
+      .compose(instance -> {
+        if (CollectionUtils.isNotEmpty(holder.getDeletedHoldingIds()) && holder.getDeletedHoldingIds().contains(holdingId)) {
```
Comment (TarasSpashchenko) [inline src/main/java/org/folio/service/orders/lines/update/instance/WithHoldingOrderLineUpdateInstanceStrategy.java:161]: **[Suggestion] Performance Optimization**

The deleted holdings check is essential for fixing the "Invalid reference" issue. Consider optimizing for performance if the number of deleted holdings could be large:

```java
// Convert deletedHoldingIds to Set for O(1) lookup instead of O(n) List.contains()
if (CollectionUtils.isNotEmpty(holder.getDeletedHoldingIds()) && 
    new HashSet<>(holder.getDeletedHoldingIds()).contains(holdingId)) {
    return Future.succeededFuture();
}
```

Or better yet, consider changing `deletedHoldingIds` in `OrderLineUpdateInstanceHolder` from `List<String>` to `Set<String>` to avoid the conversion overhead.

File: src/main/java/org/folio/service/orders/lines/update/instance/WithHoldingOrderLineUpdateInstanceStrategy.java [hunk 5]
```diff
@@ -160,8 +157,16 @@ private Future<Void> findOrCreateHoldingsAndUpdateItems(OrderLineUpdateInstanceH
...
+          return Future.succeededFuture();
+        }
+        return inventoryHoldingManager.getOrCreateHoldingRecordByInstanceAndLocation(newInstanceId, location, locationContext);
+      })
       .compose(newHoldingId -> {
+        if (Objects.isNull(newHoldingId)) {
```
Comment (TarasSpashchenko) [inline src/main/java/org/folio/service/orders/lines/update/instance/WithHoldingOrderLineUpdateInstanceStrategy.java:167]: **[Suggestion] Defensive Programming**

The null check for `newHoldingId` is excellent defensive programming. Consider adding a debug log to help with troubleshooting:

```java
if (Objects.isNull(newHoldingId)) {
    logger.debug("findOrCreateHoldingsAndUpdateItems:: newHoldingId is null, skipping item updates for holdingId: {}", holdingId);
    return Future.succeededFuture();
}
```

This would help track when holding creation/retrieval fails and processing is skipped.

File: src/test/java/org/folio/service/inventory/InventoryManagerTest.java [hunk 6]
```diff
@@ -878,6 +878,152 @@ void shouldCreateShadowInstance() {
     verify(sharingInstanceService).createShadowInstance(instanceId, configuration.get(), requestContext);
   }
 
+  @Test
```
Comment (TarasSpashchenko) [inline src/test/java/org/folio/service/inventory/InventoryManagerTest.java:881]: **[Excellent] Comprehensive Test Coverage**

Outstanding test coverage! The new test methods comprehensively cover the edge cases that were causing the "Invalid reference" issue:

1. `shouldCreateNewHoldingWhenLocationIdProvidedButNoMatchingHoldingExists` - Tests creation path
2. `shouldReturnNullWhenHoldingIdProvidedButHoldingNotFound` - Tests the null handling fix  
3. `shouldCreateNewHoldingWhenHoldingIdProvidedButNoMatchingHoldingExists` - Tests fallback creation
4. `shouldReuseExistingHoldingWhenHoldingIdProvidedAndMatchingHoldingExists` - Tests reuse logic

These tests directly validate the scenarios described in JIRA ticket MODORDSTOR-458. The test structure is clean and follows good testing practices with clear Given-When-Then structure.

**Review by TarasSpashchenko — APPROVED** [review]
## 🎯 **Code Review Summary: APPROVE**

This PR successfully addresses **JIRA ticket MODORDSTOR-458** by fixing the "Invalid reference" issue that occurred when changing instance connections for P/E mix orders. The implementation is well-architected and properly tested.

### ✅ **Key Strengths**

1. **Root Cause Resolution**: The issue was correctly identified - the system was attempting to process holdings that had already been deleted during the "Delete Holdings" operation.

2. **Clean Architecture**: The solution uses the holder pattern to centrally track deleted holdings, maintaining proper separation of concerns across the strategy classes.

3. **Defensive Programming**: Added proper null checks and guards to prevent processing of non-existent holdings.

4. **Modern Java**: Good use of modern Java syntax (`toList()`, improved collection handling).

5. **Comprehensive Testing**: Added 4 new test methods that directly validate the scenarios from the JIRA ticket, ensuring the fix works as expected.

### 🔧 **What This Fix Accomplishes**

- **Prevents "Invalid reference" display** by tracking which holdings were deleted
- **Improves user experience** for P/E mix order instance connection changes  
- **Maintains data integrity** by properly handling deleted holding references
- **Ensures workflow continuity** without breaking existing functionality

### 🎯 **Technical Implementation Quality**

- ✅ **FOLIO Architecture Compliance**: Follows established patterns and practices
- ✅ **Error Handling**: Robust null checks and graceful handling of edge cases
- ✅ **Performance**: Efficient implementation with only minor optimization suggestions
- ✅ **Security**: No security concerns identified - maintains proper data access patterns
- ✅ **Testing**: Excellent test coverage for new functionality and edge cases

### 🚀 **Recommendation**

This PR is **ready for merge**. The fix directly addresses the user-reported issue while maintaining code quality and adding proper test coverage. The changes are focused, well-implemented, and align with FOLIO's architectural principles.

**Post-merge recommendation**: Monitor the related JIRA tickets (MODORDERS-1299, MODORDERS-1300) mentioned in the ticket comments to ensure this fix doesn't impact other instance connection scenarios.
